src/arithmetic/canonical_simplify.cc

```diff
@@ -0,0 +1,830 @@
+      res = res % make_const(dtype, this->upper_factor);
+    }
+    if (this->lower_factor != 1) {
+      res = res / make_const(dtype, this->lower_factor);
+    }
+    sscale *= this->scale;
+    if (sscale != 1) {
+      CHECK(!dtype.is_uint() || sscale > 0);
+      res = res * make_const(dtype, sscale);
+    }
+    return res;
+  }
+
+  Expr Normalize() const final {
+    return NormalizeWithScale(1);
+  }
+
+  void MulToSelf(int64_t scale) {
+    this->scale *= scale;
+  }
+
+  inline bool IndexEqual(const SplitExpr& other) const;
+
+  /*! \brief positive infty */
+  static const constexpr int64_t kPosInf = ConstIntBoundNode::kPosInf;
+  static constexpr const char* _type_key = "arith.SplitExpr";
+  TVM_DECLARE_NODE_TYPE_INFO(SplitExprNode, CanonicalExprNode);
+};
+
+TVM_DEFINE_COW_NODE_REF(SplitExpr, Expr, SplitExprNode);
+
+inline bool SplitExprNode::IndexEqual(const SplitExpr& other) const {
+  if (index.same_as(other->index)) return true;
+  return ir::Equal(index, other->index);
+}
+
+/*!
+ * \brief Normal form that represents sum of expressions.
+ *
+ *  result = sum(args) + base.
+ */
+class SumExprNode : public CanonicalExprNode {
+ public:
+  /*!
+   * \brief arguments to be summed up.
+   *
+   * args are divided into segments with the same index.
+   * within each segment, the SplitExpr is ordered in descending order of lower_factor.
+   *
+   * \note Can be mutated by SimplifySplitExprs, which is idempotent
+   */
+  mutable std::vector<SplitExpr> args;
+  /*! \brief Base value in the summation. */
+  int64_t base{0};
+
+  /*! \brief Simplify the args by merging SplitExprs */
+  void SimplifySplitExprs() const {
+    // NOTE: This algorithm relies on the factor that args are divided into segments
+    // and each segment is sorted in descending order of lower_factor.
+    for (size_t i = 0; i < args.size(); ++i) {
+      if (args[i]->scale == 0) continue;
+      for (size_t j = i + 1; j < args.size(); ++j) {
+        SplitExpr& lhs = args[i];
+        SplitExpr& rhs = args[j];
+        if (!lhs->IndexEqual(rhs)) break;
+        if (lhs->upper_factor < rhs->lower_factor) break;
+        if (lhs->lower_factor == rhs->upper_factor &&
+            lhs->scale % rhs->scale == 0 &&
+            lhs->lower_factor == (lhs->scale / rhs->scale) * rhs->lower_factor) {
+          // General merge condition and proof:
+          // - x = lhs->index % lhs->upper_factor
+          // - s = lhs->scale / rhs->scale
+          // - c = rhs->lower_factor
+          //
+          //    ((x / (c * s)) * s + (x % (c * s)) / c
+          // => ((x / c) / s) * s + ((x / c) % s)
+          // => (x / c)
+          //
+          // Examples:
+          //
+          //    (z / 6) * 6 + ((z % 6) / 3) * 3
+          // => ((z / 6) * 2 + (z % 6) / 3) * 3
+          // => (z / 3) * 3
+          // note: x = z, c = 3, s = 2
+          //
+          //    ((z % 12) / 6) * 6 + ((z % 6) / 3) * 3
+          // => (((z % 12) / 6) * 2 + ((z % 12) % 6) / 3) * 3
```

Comment:
Shouldn't there be a condition that `lhs->upper_factor % rhs->upper_factor == 0` so that we can perform the transformation `z % rhs->upper_factor  =>  (z % lhs->upper_factor) % rhs->upper_factor`?

Reply:
This is an invariant condition that lhs->upper_factor % lhs->lower_factor == 0